### PR TITLE
rfc: bnorm: single-pass formula for calculating mean and variance

### PR DESCRIPTION
**Pull request by gtrzebia**
[link to rendered document](https://github.com/oneapi-src/oneDNN/tree/gtrzebia/rfc/single-pass-bnorm/rfcs/20210519-single-pass-bnorm)

### CODE REVIEW

**Review by dzarukin — COMMENTED** [review]
Hi!
I believe this RFC should answer several important questions uncovered here:
1. Who is the requester? What's the exact request is?
2. If this is about performance, what are the benefits of such approach?
3. With new formula, does it pass validation on framework side with a broad set of topologies? Is it intended for inference only or training?
4. Exact way of introducing a switch into the library with API section covered.
Could you share answers on these questions with the team, please?
Thanks.

**Comment by gtrzebia** [comment]
@dzarukin : README.md was updated to answer most of your questions. Regarding (4) I have no details and I was hoping to hear suggestions how to do it best.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Proposal to calculate mean & variance in batchnorm in single pass
+
+## Requester
+GPU Enabling team
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:4]: Usually requester is a team working on a framework or an application that requests performance optimizations or new functionality to help them to achieve/get closer to performance targets set for their application or framework.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+## Motivation
+Using new formula for calculating variance will result in up to 33% performance
+improvement in batchnorm fwd training operation. 
+
+Currently bnorm performance is far below projections. Projections are based on
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:10]: Are there any other optimizations which can be done to achieve the same performance or reduce existing gap?

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_BNORM_ONE_PASS`. If not set or equal to zero, 
+old formula will be used.
+
+Pros:
+- API/ABI compatibility
+- Simplest to implement and use
+- Clearly defined default state
+- If needed, may be kept confidential and not disclosed to external customes
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:76]: onednn is an open source project, there is no reason to keep it secret, optimizations are vendor-agnostic.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+The last line, "if var < 0 then var = 0" is there to make sure rounding errors
+won't cause variance to go negative. Further in batchnorm calculations square
+root is taken from variance. Negative value would result in NaN.
+
+New formula should be considered experimental. By default the old one will be
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_BNORM_ONE_PASS`. If not set or equal to zero,
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:69]: This doesn't sound like a production quality. We need a contract so frameworks/user/tests/examples/implementations know how to use the feature, this is exactly why we would want to add an API for the feature.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+Bnorm descriptor has flags field `dnnl_normalization_flags_t`. Add new flag 
+`dnnl_one_pass = 0x20u`. When set to non-zero, new formula will be used.
+
+Pros:
+- Fine-tuning per instance of bnorm primitive
+- Suitable to become permanent solution, should both formulas be kept in oneDNN
+
+Cons:
+- API change needed.
+- In case a customer was passing garbage into bit5 of flags, the default
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:96]: This is a user error, not a cons of a proposed solution. User also can set `DNNL_BNORM_ONE_PASS` to a garbage, so there is no differences.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+## Motivation
+Using new formula for calculating variance will result in up to 33% performance
+improvement in batchnorm fwd training operation. 
+
+Currently bnorm performance is far below projections. Projections are based on
+assumption that bnorm reads input tensor once and writes result once. Current
+implementation requires bnorm to read input tensor 3 times. Proposed change
+will decrease it to 2 reads. There is no known generic algorithm to decrease it
+to 1 read, although such efficiency is possible for special cases where input
+tensor is small enough.
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:15]: Do you know how it its implemented in other performance libraries, like cudnn? It would be helpful to understand alternative solutions

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+- Clearly defined default state
+
+Cons:
+- API change needed.
+- No fine-tuning per instance of bnorm primitive
+- If we decide to remove it later and keep only  one formula, it will require
+ another API change; it will leave behind a deprecated function.
+
+## Validation
+New formula was tested with RN50 and it trained to SotA. Further tests with
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:122]: Why do we want to have it in production if we are not sure that this algorithm works? Will it be enough for frameworks to have a separate branch with this optimization?

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+Pros:
+- API stays backwards-compatible
+- Clearly defined default state
+
+Cons:
+- API change needed.
+- No fine-tuning per instance of bnorm primitive
+- If we decide to remove it later and keep only  one formula, it will require
+ another API change; it will leave behind a deprecated function.
+
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:120]: Option 4. The feature you describe is a new algorithm for batch normalization, so I think alg_kind_t fits this feature perfectly:
```cpp
// dnnl_types.h
typedef enum {
...
dnnl_batch_normalization_direct,
dnnl_batch_normalization_one_pass, // name is still bad and doesn't tell how it is computed
...
} dnnl_alg_kind_t;

// dnnl.h
dnnl_status_t DNNL_API dnnl_batch_normalization_forward_desc_init_v2(
        dnnl_batch_normalization_desc_t *bnrm_desc, dnnl_prop_kind_t prop_kind, dnnl_alg_kind_t alg_kind,
        const dnnl_memory_desc_t *data_desc, float epsilon, unsigned flags);
```

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+## Validation
+New formula was tested with RN50 and it trained to SotA. Further tests with
+broader set of topologies are needed. Until such tests are performed, new
+formula should be optionally available but disabled by default. The point
+of this request is to make new formula available for further validation, while
+external customers keep using the old and proven formula.
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:126]: All in all I would vote for the option with an additional flag, because this way we define some API, which can be used by our tests and user. On the other side since we are not sure it will be widely used we minimize amount of changes on our side to just a single additional flag.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 4 above)
Comment (densamoilov) [reply rfcs/20210519-single-pass-bnorm/README.md:69]: Isn’t it an experimental feature?

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 2 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md:10]: There is one method to reach projection, but it is not generic.
In bnorm, each channel of data has no influence on other channels, as if they were separate tensors. If:
- tensor is small enough that one channel worth of data can all fit into cache, and
- tensor layout allows loading a single channel at a time into cache/SRAM with full bandwidth - which generally means channel is not the innermost dimension (so NHWC is bad, NCHW is fine, etc)
then bnorm can load data for given channel once and do all 3 operations on it (mean, variance, normalization) while data stays in cache. In that case using new formula gives no benefit.

If either of the above points is not met (either C is innermost dimension and channels need to be loaded more-than-one at a time to sustain mem bandwidth, or there's too much data in each channel) then a weaker version of the above optimization may be used:
- for first pass (mean&var) load data for channel/set of channels starting from beginning of tensor; later loads will evict data loaded earlier, but the tail will stay in cache
- for second pass (normalization), load data backwards, starting from the end of tensor towards beginning. That way some data will be still available in cache and won't need to be loaded from main mem again.
In such scenario it again matters whether there are 2 or 3 passes and performance.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 8 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md:120]: This would be nice if we were certain that both algorithms are going to stay. But we don't know that yet. Assuming it won't stay, then this becomes the most inconvenient solution for end users: one backwards-incompatible API change now and another one when we delete it.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 7 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md:122]: According to emails from eric.lin@intel.com, new branch, as well as build-time flags, would be inconvenient for framework people. It would be fine if we put it into, uh, certain existing non-open-source branch, but code from there gets merged to master often enough that it warrants putting it into master right now.

Comment (mgouicem) [reply rfcs/20210519-single-pass-bnorm/README.md:126]: I think the rationale to put this feature in master as experimental under environment variable is to prevent divergence between master and feature branches.

So if we have X branches with X experimental features, it is inconvenient from a user perspective to try them, or even harder to try aggregate effects of experimental features. So the idea is to put experimental features in master, hidden behind build/env flags. This way users can easily try them, and select which one they can rely on or not.

Once we collect enough feedback to deem these features production ready, we can make those flags on by default.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 8 above)
Comment (mgouicem) [reply rfcs/20210519-single-pass-bnorm/README.md:124]: Overall, it seems that it would be similar to option 2, but worse :-)
- it would require to expose new op descriptor, new dispatch, ... and as @gtrzebia mentioned, there is no way to make one default, they would just cohabit next to each other
- FWK have apparently no way to know which one to chose with that granularity, so they would just pick one algorithm, and always use the same for all BN ops.

With the two above points in mind, I don't know if the new algorithm kind or flag is worth it.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_EXPERIMENTAL_BNORM_ONE_PASS`. If not set or
+equal to zero, old formula will be used.
+
+Optionally, as a safety measure against accidentally leaving the above flag 
+in wrong state, add a `DNNL_EXPERIMENTAL` build knob (disabled by default). The
+above env var would only matter when oneDNN was built with this knob enabled.
```
[the remark this answers is not in the document]
Comment (mgouicem) [reply rfcs/20210519-single-pass-bnorm/README.md:74]: Also note that for all environment variables, we also expose a function to override/set the setting as well.

**Review by mgouicem — COMMENTED** [review]
Overall, I would favor the build/env_var option simply as it would make adoption simpler.
If we expose new flags/alg_kind, there is a high change that users will set those conservatively in code, and will not change them in the future.

Though here I don't know if we can consider this formula as a new feature, or just a relaxed math mode. So maybe the environment variable should be `DNNL_EXPERIMENTAL_FAST_MATH` or `DNNL_EXPERIMENTAL_RELAXED_MATH`

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+The last line, "if var < 0 then var = 0" is there to make sure rounding errors
+won't cause variance to go negative. Further in batchnorm calculations square
+root is taken from variance. Negative value would result in NaN.
+
+New formula should be considered experimental. By default the old one will be
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_EXPERIMENTAL_BNORM_ONE_PASS`. If not set or
```
Comment (igorsafo) [inline rfcs/20210519-single-pass-bnorm/README.md:69]:
```suggestion
OneDNN will read env var `DNNL_EXPERIMENTAL_BNORM_STATS_ONE_PASS`. If not set or
```

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 2 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md:10]: @mgouicem : that's exactly what I tried to describe in my previous comment. Indeed this is the optimal way to do bnorm, but it is possible only if tensor, or its slice of channels, is small enough to fit into registers/SLM/L3$. If tensor is too big, data will spill out and need to be reloaded.

**Comment by gtrzebia** [comment]
@karturov : "...propagated to algorithm field of VERBOSE_LOG"
do you mean this place:
dnnl_verbose,exec,gpu,batch_normalization,**ocl:gen9:blocked**,forward_training,data_f16::blocked:abcdf0 ,,flags:S,mb128ic256ih56iw56,24.666 
? If not, please point me to where you'd like it

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 4 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md]: Indeed it's experimental and build flag, env var names were updated to reflect it better.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 2 above)
Comment (mgouicem) [reply rfcs/20210519-single-pass-bnorm/README.md:10]: Sorry I had misunderstood. Makes sense.

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+- Harder to reproduce bugs related to bnorm: value of this flag will need to be
+ reported in bug reports to allow reproduction.
+ - Inconvenient long-term. State of oneDNN will have a dependency which is not
+ obvious to users.
+
+## Layernorm
+Batchnorm and layernorm primitives work according to the same principles and
+formulas. Only difference is in dimension across which the normalization is
+performed. Therefore layernorm's performance may benefit from using the new
+formula for calculating mean & variance.
```
Comment (mgouicem) [inline rfcs/20210519-single-pass-bnorm/README.md:89]: Are we sure this is useful for layermorm? In that case, statistics are computed across channels, and this is typically small enough so that the two pass algorithm should happen in register/slm.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 6 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md:15]: Intel Parallel Computing Lab/CPU, Habana/Gaudi, Nervana/Crest use faster formula.
CuDNN's batchnorm returns mean and var with imprefect precision that indicates (but not proves) they use it too.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 12 above)
Comment (gtrzebia) [reply rfcs/20210519-single-pass-bnorm/README.md:89]: It depends on tensor size. Indeed, with BERT's layernorms we encountered so far we don't need it.

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 6 above)
Comment (igorsafo) [reply rfcs/20210519-single-pass-bnorm/README.md:15]: Thanks, resolved

**Review by igorsafo — APPROVED** [review]
